src/cls/rbd/cls_rbd.cc

```diff
@@ -4872,6 +4872,190 @@ int image_get_group(cls_method_context_t hctx,
   return 0;
 }
 
+namespace trash {
+
+static const std::string IMAGE_KEY_PREFIX("id_");
+
+std::string image_key(const std::string &image_id) {
+  return IMAGE_KEY_PREFIX + image_id;
+}
+
+std::string image_id_from_key(const std::string &key) {
+  return key.substr(IMAGE_KEY_PREFIX.size());
+}
+
+} // namespace trash
+
+/**
+ * Add an image entry to the rbd trash. Creates the trash object if
+ * needed, and stores the deletion condition information associated with the
+ * image.
+ *
+ * Input:
+ * @param id the id of the image
+ * @param cond the encoded deletion condition
+ *
+ * Output:
+ * @returns -EEXIST if the image id is already in the trash
+ * @returns 0 on success, negative error code on failure
+ */
+int trash_add(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
+  int r = cls_cxx_create(hctx, false);
+  if (r < 0) {
+    CLS_ERR("could not create trash: %s", cpp_strerror(r).c_str());
+    return r;
+  }
+
+  string id;
+  cls::rbd::DeletedImageSpec del_spec;
+  try {
+    bufferlist::iterator iter = in->begin();
+    ::decode(id, iter);
+    ::decode(del_spec, iter);
+  } catch (const buffer::error &err) {
+    return -EINVAL;
+  }
+
+  if (!is_valid_id(id)) {
+    CLS_ERR("trash_add: invalid id '%s'", id.c_str());
+    return -EINVAL;
+  }
+
+  CLS_LOG(20, "trash_add id=%s", id.c_str());
+
+  string key = trash::image_key(id);
+  bufferlist tmp;
+  r = read_key(hctx, key, &tmp);
+  if (r != -ENOENT) {
+    CLS_LOG(10, "id already exists");
+    return -EEXIST;
```

Comment:
Nit: might have hit a different error from -ENOENT -- should only return -EEXISTS if "r == 0"